packages/expo-clipboard/android/src/main/java/expo/modules/clipboard/ClipboardImage.kt

```diff
@@ -0,0 +1,171 @@
+ * @return clip data ready to be shared by the [android.content.ClipboardManager]
+ */
+internal suspend fun clipDataFromBase64Image(
+  context: Context,
+  base64Image: String,
+  clipboardCacheDir: File,
+): ClipData {
+  // 1. Get bitmap from base64 string
+  val data = maybeStripBase64Prefix(base64Image)
+  val bitmap = bitmapFromBase64String(data)
+
+  // 2. Create file in cache dir, it will be overwritten if already exists
+  val file = File(clipboardCacheDir, "copied_image.jpeg").also {
+    it.ensureExists()
+  }
+
+  // 3. Write bitmap to the file
+  val fileStream = runInterruptible { FileOutputStream(file, false) }
+  BufferedOutputStream(fileStream).use { outputStream ->
+    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
+    runInterruptible { outputStream.flush() }
+  }
+
+  // 4. Get content:// URI to the image file and put it to the clipboard data
+  val imageUri = ClipboardFileProvider.getUriForFile(
+    context,
+    context.applicationInfo.packageName + ".ClipboardFileProvider",
+    file
+  )
+  return ClipData.newUri(context.contentResolver, "image", imageUri)
+}
+// endregion
+
+// region Utility functions
+
+/**
+ * Retrieves [Bitmap] from `content://` image uri
+ * @throws SecurityException when app has no permission to access the content uri
+ * @throws IOException
+ */
+internal suspend fun bitmapFromContentUriAsync(context: Context, imageUri: Uri): Bitmap =
+  runInterruptible(Dispatchers.IO) {
+    val contentResolver = context.contentResolver
+    when {
+      Build.VERSION.SDK_INT < 28 -> MediaStore.Images.Media.getBitmap(
+        contentResolver,
+        imageUri
+      )
+      else -> {
+        val source = ImageDecoder.createSource(contentResolver, imageUri)
+        ImageDecoder.decodeBitmap(source)
+      }
+    }
+  }
+
+internal fun bitmapFromBase64String(base64Image: String): Bitmap = try {
+  val byteArray = Base64.decode(base64Image, Base64.DEFAULT)
+  BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
+    ?: throw RuntimeException("Failed to convert base64 into Bitmap")
+} catch (e: RuntimeException) {
+  // Base64.decode throws IllegalArgumentException on invalid data, but
+  // BitmapFactory.decodeByteArray returns null if data cannot be converted
+  // so we aggregate both errors here
+  throw InvalidImageException(base64Image, e)
+}
+
+/**
+ * Strips the `data:image/...;base64.` prefix from the base64 string
+ * leaving the pure base64 data string.
+ * Does nothing if there's no such prefix.
+ */
+private fun maybeStripBase64Prefix(base64String: String): String =
+  if (base64String.startsWith("data:")) {
+    val lastPrefixCharPos = base64String.indexOf(',')
+    base64String.substring(lastPrefixCharPos + 1)
+  } else {
+    base64String
+  }
+
+/**
+ * Creates the file and all its parent directories if they don't exist already.
+ */
+private suspend fun File.ensureExists() {
+  parentFile?.mkdirs()
+  runInterruptible(Dispatchers.IO) { createNewFile() }
```

Comment:
Wouldn't `parentFile?.mkdirs()` eventually end in performing `createNewFile()` somewhere? 🤔 
Might that cause race condition if not run on the same scope?
```suggestion
  runInterruptible(Dispatchers.IO) {  
    parentFile?.mkdirs()
    createNewFile()
  }
```

Reply:
- `somePath.mkdirs()` treats `somePath` as directory, creates it and its parent directories
- `someFile.createNewFile()` treats `someFile` as file and creates an empty file (if not exists). Throws if its parent directory doesn't exist (so mkdirs must be called first on parent)
- `runInterruptible` should be called only on Java blocking calls, which support `InterruptedException`. But maybe putting `mkdirs` inside isn't wrong. Need to research it more.

Reply:
Ok, I've haven't communicated my question the right way 😅 
`parentFile?.mkdirs()` might end up performing some `createNewDirectory()` that would internally call `createNewFile(isDirectory=true)`.
My concern, that might be completely inaccurate, is:
- is `parentFile?.mkdirs()` a synchronous call or it's rather asynchronous and you're not waiting for the result here? 🤔
- because if it's asynchronous and the code flow goes on then it would trigger another fileSystem function `createNewFile()` (that is running on different thread) and this `createNewFile()` might end being called when it's parent directory is still missing.
- I guess I'm trying to understand how the coroutines are performing their job under the hood and if there's no fs-related race condition happening here 😅

Reply:
They're both synchronous 😄 so the `runInterruptible` will start _after_ `mkdirs` finishes. The purpose of `runInterruptible` here is different